core/src/main/java/com/linecorp/armeria/client/websocket/WebSocketSession.java

```diff
@@ -0,0 +1,141 @@
+import com.linecorp.armeria.common.websocket.WebSocket;
+import com.linecorp.armeria.common.websocket.WebSocketFrame;
+import com.linecorp.armeria.common.websocket.WebSocketWriter;
+import com.linecorp.armeria.internal.common.websocket.WebSocketFrameEncoder;
+
+/**
+ * A WebSocket session that is created after {@link WebSocketClient#connect(String)} succeeds.
+ * You can start sending {@link WebSocketFrame}s via {@link #setOutbound(Publisher)}. You can also subscribe to
+ * {@link #inbound()} to receive {@link WebSocketFrame}s from the server.
+ */
+@UnstableApi
+public final class WebSocketSession {
+
+    private final ClientRequestContext ctx;
+    private final ResponseHeaders responseHeaders;
+    @Nullable
+    private final String subprotocol;
+    private final WebSocket inbound;
+    private final CompletableFuture<StreamMessage<HttpData>> outboundFuture;
+    private final WebSocketFrameEncoder encoder;
+
+    WebSocketSession(ClientRequestContext ctx, ResponseHeaders responseHeaders, WebSocket inbound,
+                     CompletableFuture<StreamMessage<HttpData>> outboundFuture,
+                     WebSocketFrameEncoder encoder) {
+        this.ctx = ctx;
+        this.responseHeaders = responseHeaders;
+        subprotocol = responseHeaders.get(HttpHeaderNames.SEC_WEBSOCKET_PROTOCOL);
+        this.inbound = inbound;
+        this.outboundFuture = outboundFuture;
+        this.encoder = encoder;
+    }
+
+    /**
+     * Returns the {@link ClientRequestContext}.
+     */
+    public ClientRequestContext context() {
+        return ctx;
+    }
+
+    /**
+     * Returns the {@link ResponseHeaders}.
+     */
+    public ResponseHeaders responseHeaders() {
+        return responseHeaders;
+    }
+
+    /**
+     * Returns the subprotocol negotiated between the client and the server.
+     */
+    @Nullable
+    public String subprotocol() {
+        return subprotocol;
+    }
+
+    /**
+     * Returns the {@link WebSocket} that is used to receive WebSocket frames from the server.
+     */
+    public WebSocket inbound() {
+        return inbound;
+    }
+
+    /**
+     * Returns the {@link WebSocketWriter} that is used to send WebSocket frames to the server.
+     *
+     * @throws IllegalStateException if this method or {@link #setOutbound(Publisher)} has been called already.
+     */
+    public WebSocketWriter outbound() {
+        final WebSocketWriter writer = WebSocket.streaming();
+        setOutbound(writer);
+        return writer;
+    }
+
+    /**
+     * Sets the {@link WebSocket} that is used to send WebSocket frames to the server.
+     *
+     * @throws IllegalStateException if this method or {@link #outbound()} has been called already.
+     */
+    public void setOutbound(Publisher<? extends WebSocketFrame> outbound) {
+        requireNonNull(outbound, "outbound");
+        if (outboundFuture.isDone()) {
+            throw new IllegalStateException("outbound() or setOutbound() has been already called.");
```

Comment:
Should we call `outbound.abort()` if `outbound` is an instance of `StreamMessage`?

Reply:
Yeah, we need it. Thanks! 😉